src/clang-cpp-frontend/clang_cpp_convert.cpp

```diff
@@ -1184,62 +1193,97 @@ bool clang_cpp_convertert::get_function_body(
+           * initializer can legally expect previous members obtained via the
+           * (this) pointer to have already been initialized. */
           const exprt &this_ptr = member.op0(); /* the (this) pointer */
           const struct_union_typet &this_type =
             to_struct_union_type(ns.follow(this_ptr.type().subtype()));
-          bool is_struct = this_type.is_struct();
-          exprt rhs_sym(is_struct ? "struct" : "union", this_type);
-          if (is_struct)
+
+          if (!array_init_sym)
           {
-            for (const struct_union_typet::componentt &c :
-                 this_type.components())
-              if (c.get_name() == member.component_name())
-                rhs_sym.move_to_operands(rhs);
-              else
-              {
-                member_exprt old(this_ptr, c.get_name(), c.type());
-                rhs_sym.move_to_operands(old);
-              }
+            symbolt new_symbol;
+            new_symbol.name = "array_init$";
+            new_symbol.id = id2string(this_ptr.identifier()) + "_array_init$";
+            new_symbol.type = this_type;
+            if (context.move(new_symbol, array_init_sym))
+            {
+              log_error(
+                "Duplicate temporary array symbol for initializer in {}",
+                __func__);
+              abort();
+            }
+            assert(array_init_sym);
           }
-          else
+
+          exprt this_sym = dereference_exprt(this_ptr, this_ptr.type());
+          exprt init_sym = symbol_expr(*array_init_sym);
+
+          if (!init_sym_uptodate)
           {
-            /* single initializer for unions */
-            rhs_sym.component_name(member.component_name());
-            rhs_sym.move_to_operands(rhs);
+            /* 0. array_init$ = *this; */
+            side_effect_exprt update("assign", this_type);
+            update.copy_to_operands(init_sym, this_sym);
+            initializers.push_back(update);
+            init_sym_uptodate = true;
           }
 
-          exprt this_sym = dereference_exprt(this_ptr, this_type);
+          exprt new_member = member;
+          new_member.op0() = init_sym;
+
+          /* 1. array_init$.member = rhs; */
+          initializer = side_effect_exprt("assign", new_member.type());
+          initializer.move_to_operands(new_member, rhs);
+          initializer.location() = new_expr.location();
+          initializers.push_back(initializer);
+
+          /* 2. *this = array_init$; */
           initializer = side_effect_exprt("assign", this_sym.type());
-          initializer.move_to_operands(this_sym, rhs_sym);
+          initializer.move_to_operands(this_sym, init_sym);
+          initializer.location() = new_expr.location();
+          initializers.push_back(initializer);
         }
         else
         {
           initializer = side_effect_exprt("assign", member.type());
+          initializer.set("#member_init", 1);
```

Comment:
Do we need it?

Reply:
No, I don't think so. I forgot to remove it, sorry.